src/value/use-viewport-scroll.ts

```diff
@@ -7,55 +10,88 @@ export interface ScrollMotionValues {
     scrollYProgress: MotionValue<number>
 }
 
-const scrollX = motionValue(0)
-const scrollY = motionValue(0)
-const scrollXProgress = motionValue(0)
-const scrollYProgress = motionValue(0)
-
-const setProgress = (offset: number, maxOffset: number, value: MotionValue) => {
-    value.set(!maxOffset || !offset ? 0 : offset / maxOffset)
+function getWindowOffsets() {
+    return {
+        xOffset: window.pageXOffset,
+        yOffset: window.pageYOffset,
+        xMaxOffset: document.body.clientWidth - window.innerWidth,
+        yMaxOffset: document.body.clientHeight - window.innerHeight,
+    }
 }
 
-let hasEventListener = false
-const addScrollListener = () => {
-    hasEventListener = true
-
-    if (typeof window === "undefined") return
-
-    const updateScrollValues = () => {
-        const xOffset = window.pageXOffset
-        const yOffset = window.pageYOffset
-
-        // Set absolute positions
-        scrollX.set(xOffset)
-        scrollY.set(yOffset)
-
-        // Set 0-1 progress
-        setProgress(
-            xOffset,
-            document.body.clientWidth - window.innerWidth,
-            scrollXProgress
-        )
-        setProgress(
-            yOffset,
-            document.body.clientHeight - window.innerHeight,
-            scrollYProgress
-        )
+function getElementOffsets(element: HTMLElement) {
+    return {
+        xOffset: element.scrollLeft,
+        yOffset: element.scrollTop,
+        xMaxOffset: element.scrollWidth - element.offsetWidth,
+        yMaxOffset: element.scrollHeight - element.offsetHeight,
     }
+}
 
-    updateScrollValues()
+function createWindowListeners(updateCallback) {
+    window.addEventListener("resize", updateCallback)
+    window.addEventListener("scroll", updateCallback, { passive: true })
+    return () => {
+        window.removeEventListener("resize", updateCallback)
+        window.removeEventListener("scroll", updateCallback, { passive: true })
+    }
+}
 
-    window.addEventListener("resize", updateScrollValues)
-    window.addEventListener("scroll", updateScrollValues, { passive: true })
+function createElementListeners(element: HTMLElement, updateCallback) {
+    let resizeObserver
+    if (window.ResizeObserver) {
```

Comment:
`ResizeObserver` has very poor support (around 60%). I don't know if it is library containers or consumers who should supply polyfill for that.